### PR TITLE
fmt: pickup v9.1.0

### PR DESCRIPTION
**Pull request by tchaikov**
fmt 9.0.0 dropped automatic `std::ostream` insertion operator discovery when `fmt/ostream.h` to prevent ODR violations. instead of defining `FMT_DEPRECATED_OSTREAM`, we took efforts to specialize `fmt::formatter<..>` to be compatible with the new fmtlib. to avoid breaking the build with fmt v9 and up, let's bump up the fmt submodule.

Signed-off-by: Kefu Chai <tchaikov@gmail.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by tchaikov** [comment]
marking as a draft as it fails to link.

File: src/rgw/rgw_common.h [hunk 1]
```diff
@@ -20,6 +20,11 @@
 #include <string_view>
 #include <atomic>
 #include <unordered_map>
+#include <fmt/format.h>
+
+#undef FMT_HEADER_ONLY
+#define FMT_HEADER_ONLY 1
+#include <fmt/format.h>
```
Comment (cbodley) [inline src/rgw/rgw_common.h:27]: was that first `#include` line intentional?

Comment (tchaikov) [reply src/rgw/rgw_common.h]: @cbodley it's not. thanks for pointing this out. i corrected it in the latest revision. hopefully it is able to address link failure.

**Comment by tchaikov** [comment]
jenkins retest this please.

**Comment by tchaikov** [comment]
@cbodley hi Casey, could you help re-review this PR?

**Comment by tserong** [comment]
Code looks fine to me. I've just kicked off a build on openSUSE, will report back once complete.

**Comment by tchaikov** [comment]
@tserong hi Tim, thank you in advance!

**Review by tserong — APPROVED** [review]
LGTM, and built fine on openSUSE Tumbleweed

File: src/rgw/rgw_common.h [hunk 2]
```diff
@@ -1654,6 +1656,13 @@ inline std::ostream& operator<<(std::ostream& out, const rgw_obj_key &o) {
   return out << o.to_str();
 }
 
+template<> struct fmt::formatter<rgw_obj_key> : fmt::formatter<std::string_view> {
+  template <typename FormatContext>
+  auto format(const rgw_obj_key& key, FormatContext& ctx) const {
+    return formatter<std::string_view>::format(key.to_str(), ctx);
```
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/rgw/rgw_common.h:1662]: no, AFAICT. `formatter<std::string_view>` is quite popular when implementing a customized formatter for its formatter modifier support. when it comes to the life cycle of the temporary `std::string`, C++ standard requires the temporary object to be alive when evaluating the expression using it. in this case the expression is `formatter<std::string_view>::format(key.to_str(), ctx)`, where we also implicitly call the `operator std::string_view()` of `std::string`. a minimal sample can be found at https://godbolt.org/z/d647sYbd3.

[the remark this answers is not in the document]
Comment (tchaikov) [reply src/rgw/rgw_common.h]: anyway, i moved the implementation of `rgw_obj_key::to_str()` into `fmt::formatter<rgw_obj_key>` for better readability.

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
jenkins test api

**Comment by tchaikov** [comment]
jenkins test api

File: src/rgw/rgw_common.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (cbodley) [reply src/rgw/rgw_common.h]: thanks!

**Review by cbodley — APPROVED** [review]

File: src/rgw/rgw_common.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (dang) [reply src/rgw/rgw_common.h]: This broke the Motr store, which made extensive use of to_str().   What should replace to_str() usage for getting a string (not for operator<< usage)?

[the remark this answers is not in the document]
Comment (cbodley) [reply src/rgw/rgw_common.h]: (but feel free to add `to_str()` back just to make it compile)

File: src/rgw/driver/rados/rgw_data_sync.h [hunk 3]
```diff
@@ -340,9 +338,9 @@ struct RGWDataSyncEnv {
 
 // pretty ostream output for `radosgw-admin bucket sync run`
 template<typename ...T>
-void pretty_print(const RGWDataSyncEnv* env, T&& ...t) {
+void pretty_print(const RGWDataSyncEnv* env, fmt::format_string<T...> fmt, T&& ...t) {
   if (unlikely(!!env->ostr)) {
-    fmt::print(*env->ostr, std::forward<T>(t)...);
+    fmt::print(*env->ostr, fmt, std::forward<T>(t)...);
```
Comment (cbodley) [inline src/rgw/driver/rados/rgw_data_sync.h:343]: this doesn't seem to work with fmt-devel 8.1.1 (fedora 36):
```
In file included from ceph/src/rgw/driver/rados/rgw_data_sync.h:8,
                 from ceph/src/rgw/driver/rados/rgw_data_sync.cc:13:
/usr/include/fmt/ostream.h: In instantiation of ‘void fmt::v8::print(std::basic_ostream<Char>&, const S&, Args&& ...) [with S = basic_format_string<char, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&>; Args = {std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&}; Char = char]’:
ceph/src/rgw/driver/rados/rgw_data_sync.h:343:15:   required from ‘void pretty_print(const RGWDataSyncEnv*, fmt::v8::format_string<T ...>, T&& ...) [with T = {std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&}; fmt::v8::format_string<T ...> = fmt::v8::basic_format_string<char, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&>]’
ceph/src/rgw/driver/rados/rgw_data_sync.cc:5779:17:   required from here
/usr/include/fmt/ostream.h:130:28: error: no matching function for call to ‘to_string_view(const fmt::v8::basic_format_string<char, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&>&)’
  130 |   vprint(os, to_string_view(format_str),
      |              ~~~~~~~~~~~~~~^~~~~~~~~~~~
```
full log https://paste.centos.org/view/raw/3229396b

Comment (cbodley) [reply src/rgw/driver/rados/rgw_data_sync.h:343]: the error seems to be related to the rgw_obj_key ostream operator

Comment (tchaikov) [reply src/rgw/driver/rados/rgw_data_sync.h:343]: @cbodley #49479 should fix this. sorry for the inconvenience.

**Review by idryomov — COMMENTED** [review]
`WITH_FMT_VERSION` wasn't updated here and in a couple of previous bumps: it's still set to `7.0.0`.  Is the mismatch between the bundled version and the requirement for a system-provided package intended (e.g. to continue supporting compiling with older versions of the library)?

I also see that `run-make.sh` (used by `run-make-check.sh` which drives the `make check` job) passes `-DWITH_FMT_VERSION=9.0.0`, so we are really inconsistent here.

**Comment by cbodley** [comment]
@idryomov personally, i would love to drop support for libfmt versions earlier than 9, along with all instances of `#if FMT_VERSION >= 90000` that were needed to work around v9 API changes. we're bundling v9 anyway, so there isn't much benefit to supporting older system packages

**Comment by tchaikov** [comment]
> `WITH_FMT_VERSION` wasn't updated here and in a couple of previous bumps: it's still set to `7.0.0`. Is the mismatch between the bundled version and the requirement for a system-provided package intended (e.g. to continue supporting compiling with older versions of the library)?
> 

it is intended. lemme quote part of the cover letter here:

>> to avoid breaking the build with fmt v9 and up, let's bump up the fmt submodule

the goal was to bump the minimal fmtlib version supported by Ceph, but to avoid breaking the build with fmtlib v9 and up.

the reason why i didn't update the minimal fmtlib version supported is to keep the backward compatibility so that downstream package maintainers of various distros can build and link against the fmtlib libraries when building Ceph.

**Comment by tchaikov** [comment]
see also https://github.com/ceph/ceph/pull/48705#issuecomment-1647515897

**Comment by idryomov** [comment]
> the reason why i didn't update the minimal fmtlib version supported is to keep the backward compatibility so that downstream package maintainers of various distros can build and link against the fmtlib libraries when building Ceph. so far, quite few of popular distros have not packaged fmt v9.1.0 yet.

I think that needs to be balanced against the build/test matrix and general maintainability concerns.  For example, is carrying ~90 `ifdef`s for something as simple as formatting worth it (noting that `fmt` is used only in Crimson and a few random places -- if all of Ceph used it, it would have been many hundreds if not a thousand)?  But thanks for the confirmation!

@tchaikov If you have a moment, I would appreciate if you could take a peak at https://github.com/ceph/ceph/pull/51197/files#diff-148715d6ea0c0ea0a346af3f6bd610d010d490eca35ac6a9b408748f7ca9e3f4.  This is a change to respect `WITH_FMT_HEADER_ONLY` when using the bundled version.  I was surprised to learn that that wasn't the already the case.

**Pull request merged**